python/tank/context.py

```diff
@@ -624,37 +623,105 @@ def _fields_from_template_tree(self, template, fields, entities):
+                    #    Type     | Id  | Name     | Path
+                    #    ----------------------------------------------------
+                    #    Sequence | 001 | Seq_001  | /Seq_001
+                    #    Shot     | 002 | Shot_A   | /Seq_001/Shot_A
+                    #    Step     | 003 | Lighting | /Seq_001/Shot_A/Lighting
+                    #    Step     | 003 | Lighting | /Seq_001/blah/Shot_B/Lighting   <- this is out of date!
+                    #    Shot     | 004 | Shot_B   | /Seq_001/blah/Shot_B            <- this is out of date!
+                    #
+                    # (Note: the schema/templates have been changed since the entries for Shot_b were added)
+                    #
+                    # The sub-templates used to search for fields are:
+                    #    /{Sequence}
+                    #    /{Sequence}/{Shot}
+                    #    /{Sequence}/{Shot}/{Step}
+                    #
+                    # And the entities passed into the method are:
+                    #    Sequence:   Seq_001
+                    #    Shot:       Shot_B
+                    #    Step:       Lighting
+                    #
+                    # We are searching for fields for 'Shot_B' that has a broken entry in the path cache so the fields 
+                    # returned for each level of the template will be:
+                    #    /{Sequence}                 -> {"Sequence":"Seq_001"} <- Correct
+                    #    /{Sequence}/{Shot}          -> {}                     <- entry not found for Shot_B matching 
+                    #                                                             the template
+                    #    /{Sequence}/{Shot}/{Step}   -> {"Sequence":"Seq_001", <- Correct
+                    #                                    "Shot":"Shot_A",      <- Wrong!
+                    #                                    "Step":"Lighting"}    <- Correct
+                    #
+                    # The final fields returned are:
+                    #
+                    #     {"Sequence":"Seq_001",
+                    #      "Shot":"Shot_A",
+                    #      "Step":"Lighting"}
+                    #
+                    # The wrong Shot (Shot_A) is returned and not caught because the code only tested that the Step
+                    # entity matches and just assumes that the rest is correct - this isn't the case when there is
+                    # a one-to-many relationship between entities!
+                    #
+                    # Therefore, we need to validate that we didn't find any entity fields that we should have found
+                    # previously/higher up in the template definition.  If we did then the entries that were found 
+                    # may not be correct so we have to discard them!
+                    found_mismatching_field = False
+                    for field_name, field_value in entity_fields.iteritems():
```

Comment:
not a big deal, but are we testing the same things over and over again during the looping?